DOC: Updating capitalization of  doc/source/development

cleconte987 opened the pull request:
Regarding issue #32550. Changes to documentation folder doc/source/development to capitalise title strings and keep keyword exceptions as is

cleconte987 committed:
Updating capitalization in folder doc/source/reference

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into changes_to_pandas_remote

cleconte987 committed:
Add folder doc/source/reference in ci code checking  in ci/code_checks.sh L 330

cleconte987 committed:
Modification of validate_rst_title_capitalization.py script

cleconte987 committed:
Modify validate_rst_title_capitalization.py script to take into account words with multiple capital letters

cleconte987 committed:
remove stylistic errors in comment section L:77,78,79

cleconte987 committed:
Modify validate_rst_title_capitalization.py script

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into script_modification
Merging new changes from pandas-dev/pandas repository into script_modification to be up-to-date

cleconte987 committed:
modify script following tests

cleconte987 committed:
Modify second time following style checking

cleconte987 committed:
Update branch with pandas repository commits
Merge remote-tracking branch 'upstream_main_pandas/master' into script_modification

cleconte987 committed:
Commit to update capitalization in folder doc/source/reference and to add elements to exceptions's set in scripts/validate_rst_title_capitalization.py

cleconte987 committed:
Merge branches to update PR
Merge branch 'script_modification' into changes_to_pandas_remote

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into script_modification

cleconte987 committed:
Merge remote-tracking branch 'temporary_repo/changes_to_pandas_remote' into changes_to_pandas_remote

cleconte987 committed:
Remove wrong lines of code

cleconte987 committed:
Formatting

cleconte987 committed:
Update branch
Merge remote-tracking branch 'upstream_main_pandas/master' into changes_to_pandas_remote

cleconte987 committed:
Remove exception "GroupBy" and lower "GroupBy" in doc

cleconte987 committed:
modifying for PR

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into changes_to_pandas_remote

cleconte987 committed:
Update "groupby" to "GroupBy"

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into changes_to_pandas_remote

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into changes_to_documentation

cleconte987 committed:
Commit changes to folder doc/source/development

cleconte987 committed:
Add keywords to CAPITALIZATION_EXCEPTIONS

MomIsBestFriend commented on doc/source/development/code_style.rst:
```diff
@@ -18,7 +18,7 @@ consistent code format throughout the project. For details see the
 ========
 
-foo.__class__
+Foo.class
```
Actually this is on purpose, does ```validate_rst_title_capitalization.py``` showing this as an error?

cleconte987 replied:
It is showing this as an error, because foo is not inside exceptions.
But moreover, it is removing underscores from the title here:
`with open(rst_file, "r") as fd:
        previous_line = ""
        for i, line in enumerate(fd):
            line = line[:-1]
            line_chars = set(line)
            if (
                len(line_chars) == 1
                and line_chars.pop() in symbols
                and len(line) == len(previous_line)
            ):
                **yield re.sub(r"[{backtick}\*_]", "", previous_line)**, i
            previous_line = line`

datapythonista replied:
@cleconte987 can you change the title to `Using foo.__class__` instead? And below same for `Using f-strings`.

This will make the validation pass, and things look probably even better, without adding extra complexity.

Thanks!

cleconte987 replied:
Roger that!

MomIsBestFriend commented on doc/source/development/extending.rst:
```diff
@@ -95,7 +95,7 @@ on :ref:`ecosystem.extensions`.
 The interface consists of two classes.
 
-:class:`~pandas.api.extensions.ExtensionDtype`
+Class:~pandas.API.extensions.ExtensionDtype
```
`:class:` is something that we use for sphinx, I don't think that this is something we want

datapythonista commented:
Thanks @cleconte987. Most of this is duplicated of #32944, we're going to merge that one, so you may want to revert those files. Also, try to coordinate with @themien who is also working on this.

For the titles that are like `:class:...` we can't apply your changes, that is a special encoding to automatically create links. What we'll have to do is to update the script to skip titles that start with `:`.

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into changes_to_documentation

cleconte987 commented:
> Thanks @cleconte987. Most of this is duplicated of #32944, we're going to merge that one, so you may want to revert those files. Also, try to coordinate with @themien who is also working on this.
> 
> For the titles that are like `:class:...` we can't apply your changes, that is a special encoding to automatically create links. What we'll have to do is to update the script to skip titles that start with `:`.

Ok I will follow what @themien is doing.
And I update PR soon

cleconte987 committed:
Commit and updating documentation syntax and capitalization for folder doc/source/development

Updating script validate_rst_title_capitalization.py to avoid taking into account titles beginning with ':'

cleconte987 committed:
Remove the 'ExtensionDtype' and 'ExtensionArray' exceptions from 'CAPITALIZATION_EXCEPTIONS' that are present in title with special encoding, as it is not necessary anymore

cleconte987 committed:
Style formatting

MomIsBestFriend commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -233,6 +232,8 @@ def main(source_paths: List[str], output_format: str) -> bool:
 
     for filename in find_rst_files(source_paths):
         for title, line_number in find_titles(filename):
+            if ":class" in title:
+                print(title)
```
I *think* this should be:

```python
if ":class" in title:
   continue
```

cleconte987 replied:
Horrible... I forgot to remove it. Correcting it right away!

cleconte987 replied:
Hmm Im not so much used to git yet.. Apart from that, I didn't roll back previous changes that I made, I thought it will be ok. I don't know if the changes i didn't remove will fail

cleconte987 replied:
Or if it conflicts with previous PR or something

cleconte987 committed:
Correct error by removing wrong lines of code in validate_rst_title_capitalization.py

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -121,6 +175,7 @@ def find_titles(rst_file: str) -> Generator[Tuple[str, int], None, None]:
                 len(line_chars) == 1
                 and line_chars.pop() in symbols
                 and len(line) == len(previous_line)
+                and previous_line[0] != ":"
```
Sorry, I think I suggested this approach myself, but I see a problem now. This will skip the whole line from validation, and I see we use a mix of these `:class:` directives and regular text in some titles.

Instead of skipping the line here, I think we should update the condition in line 139. So, we check that `word` is not one of the exceptions, and also, it doesn't start with `:`. We could use a regex probably, but let's start simple, and see if it's enough.

What do you think?

cleconte987 replied:
The problem is that:

- First of all this part of code line 180

    > yield re.sub(r"[`\*_]", "", previous_line), i

    removes "`" from the title. So when function checks if title is correct, it will already be modified. So this is the main problem.

- Secondly, line 129:

    `correct_title: str = re.sub(r"^\W*", "", title).capitalize()`

    removes all non word characters at the beginning from the desired correct title that we want to keep, so ":" is not kept as being part of the correct title

- Thirdly, this line, number 136

    `word_list = re.split(r"\W", removed_https_title)`

    removes all non word characters from the list to analyse. So ":" don't appear anymore.

By the way, code in line 180 causes problems because it also removes "_" which is present in documentation like > code_style.rst line 21

`Using foo.__class__`

So what I suggest is to modify in script "validate_rst_title_capitalization.py" from line 180 on:

NB: I replace "`" by {backtick} as it causes trouble to display it as code

`yield re.sub(r"[{backtick}\*_]", "", previous_line), i`

to:

`if previous_line[0] != ":": yield re.sub(r"[{backtick}\*_]", "", previous_line), i
else: yield re.sub(r"[\*_]", "", previous_line), i`

in order to keep the "`" if title begins by a ":"
And to add in "validate_rst_title_capitalization.py" line 126:

`if title[0] == ":": return title`

So that if title begins by ":" it considers title to be valid no matter what.

Or to simply add in "validate_rst_title_capitalization.py" line 126:

`if title[0] == ":": return title`

for the same reason.


Note that there is still a problem with:

> yield re.sub(r"[`\*_]", "", previous_line), i

That doesn't keep "_" in the titles

datapythonista reviewed (approved):
Ok, let's fix it this way then, and we can find a better solution in the future.

Thanks for the work on this @cleconte987

simonjayhawkins commented:
@cleconte987 can you merge upstream master into your branch to resolve merge conflicts. see https://pandas.pydata.org/docs/development/contributing.html#updating-your-pull-request

cleconte987 committed:
Weird behaviour file 'pandas/tests/frame/indexing/test_setitem.py' not tracked anymore? Add it to repository

cleconte987 committed:
merging remote pandas-dev/pandas/master

cleconte987 committed:
Commit  modifying script 'validate_rst_title_capitalization.py' to change the way code excludes specific title syntax from being seen as an error

pep8speaks commented:
Hello @cleconte987! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`scripts/validate_rst_title_capitalization.py`](https://github.com/pandas-dev/pandas/blob/5fc4dcc70cfc9a8e3f34c732babc4f1ee844dca9/scripts/validate_rst_title_capitalization.py):

> [Line 126:89](https://github.com/pandas-dev/pandas/blob/5fc4dcc70cfc9a8e3f34c732babc4f1ee844dca9/scripts/validate_rst_title_capitalization.py#L126): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 88 characters)

cleconte987 committed:
Correct style errors

cleconte987 committed:
Merge remote-tracking branch 'upstream_main_pandas/master' into changes_to_documentation

datapythonista commented on ci/code_checks.sh:
```diff
@@ -327,7 +327,7 @@ if [[ -z "$CHECK" || "$CHECK" == "docstrings" ]]; then
 
     MSG='Validate correct capitalization among titles in documentation' ; echo $MSG
-    $BASE_DIR/scripts/validate_rst_title_capitalization.py $BASE_DIR/doc/source/development/contributing.rst
+    $BASE_DIR/scripts/validate_rst_title_capitalization.py $BASE_DIR/doc/source/development/contributing.rst $BASE_DIR/doc/source/reference
```
If everything in `development/` is fixed, can you change this in a new PR, so the whole directory is validated in the CI? Thanks!
```suggestion
    $BASE_DIR/scripts/validate_rst_title_capitalization.py $BASE_DIR/doc/source/development $BASE_DIR/doc/source/reference
```

cleconte987 replied:
Sure!

datapythonista commented:
Thanks for working on this @cleconte987

cleconte987 commented:
You're welcome!
